src/wasm2.rs

```diff
@@ -0,0 +1,222 @@
+#![allow(dead_code, clippy::panic, clippy::unwrap_used)]
+use std::sync::{Arc, Mutex};
+
+use futures::prelude::*;
+use wasm_bindgen::prelude::*;
+use wasm_bindgen_futures::future_to_promise;
+
+use crate::LspServer;
+
+/// Initialize logging - this requires the "console_log" feature to function,
+/// as this library adds 180k to the wasm binary being shipped.
+#[allow(non_snake_case, dead_code)]
+#[wasm_bindgen]
+pub fn initLog() {
+    #[cfg(feature = "console_log")]
+    console_log::init_with_level(log::Level::Trace)
+        .expect("error initializing log");
+}
+
+struct Incoming {
+    messages: Arc<Mutex<Vec<String>>>,
+}
+
+impl futures::io::AsyncRead for Incoming {
+    #[inline]
+    fn poll_read(
+        self: std::pin::Pin<&mut Self>,
+        _: &mut async_std::task::Context<'_>,
+        buffer: &mut [u8],
+    ) -> async_std::task::Poll<async_std::io::Result<usize>> {
+        let mut messages = self.messages.lock().unwrap();
+        if messages.is_empty() {
+            return async_std::task::Poll::Pending;
+        }
+        let mut byte_count = 0;
+        for message in messages.iter() {
+            log::debug!("writing message for read: {:?}", message);
+            let length = message.len();
+            buffer[byte_count..length]
```

Comment:
We need to check that `buffer` actually has room for the messages, this will panic if the message (or the sum of the messages) are larger than `buffer`.

Reply:
Yeah, there are a number of issues still left out in the I/O area. I would like to merge this patch as-is, and begin another phase of the work.